src/authentication/DPoPWebIdExtractor.ts

```diff
@@ -0,0 +1,43 @@
+import { verify } from 'ts-dpop';
+import type { TargetExtractor } from '../ldp/http/TargetExtractor';
+import { getLoggerFor } from '../logging/LogUtil';
+import type { HttpRequest } from '../server/HttpRequest';
+import type { Credentials } from './Credentials';
+import { CredentialsExtractor } from './CredentialsExtractor';
+
+/**
+ * Credentials extractor which extracts a WebID from a DPoP token.
+ */
+export class DPoPWebIdExtractor extends CredentialsExtractor {
+  protected readonly logger = getLoggerFor(this);
+  private readonly targetExtractor: TargetExtractor;
+
+  public constructor(targetExtractor: TargetExtractor) {
+    super();
+    this.targetExtractor = targetExtractor;
+  }
+
+  public async canHandle({ headers }: HttpRequest): Promise<void> {
+    const { authorization, dpop } = headers;
+    if (!authorization || !authorization.startsWith('DPoP ')) {
+      throw new Error('No DPoP Authorization header specified.');
+    }
+    if (!dpop) {
+      throw new Error('No DPoP token specified.');
+    }
+  }
+
+  public async handle(request: HttpRequest): Promise<Credentials> {
+    const { headers: { authorization, dpop }, method } = request;
+    const resource = await this.targetExtractor.handleSafe(request);
+    try {
+      const webID = await verify(authorization as string, dpop as string, method as any, resource.path);
+      this.logger.info(`Verified WebID via DPoP token: ${webID}`);
+      return { webID };
+    } catch (error: unknown) {
+      const message = `Error verifying WebID via DPoP token: ${(error as Error).message}`;
+      this.logger.warn(message);
+      throw new Error(message);
```

Comment:
Going to write a comment here to cover #361 as I think it is something that should be fixed in this PR. This is also related to the two other comments I made about errors.

The reason a 500 gets thrown as said in #361 is because at least one of the credentials extractors throws a 500 (and a composite handler takes the highest range of all the errors that gets thrown as seen [here](https://github.com/solid/community-server/blob/master/src/util/FirstCompositeHandler.ts#L100-L109).

The other 2 definitely shouldn't throw a 500 (normal errors get converted to 500, but we also have an `InternalServerError` so that one should be used if an actual 500 happens for clarity). So the question remains if a 500 should be thrown here. I don't know enough about what happens in the ts-dpop library so I don't know if there are cases when it should be a 400 and other cases when it should be a 500. If that is the case, is there any way to differentiate them based on the error that gets thrown? Even easier would be if we can always assume it's a 400 of course.

Reply:
Fixed it, thanks for the inspiration.